### PR TITLE
(#99) fix: change status code twice or with invalid code would panic now

### PR DESCRIPTION
**Pull request by dannypsnl**

### CODE REVIEW

**Comment by dannypsnl** [comment]
@shana0440 What do you think about this fail, it causes by standard lib, basically I still want support 1.9  this year.

**Comment by dannypsnl** [comment]
The question is, do you know any way could separate implementation for different Go version?

**Comment by dannypsnl** [comment]
I think I have to write some test at night, so don't merge this until I complete my job

**Comment by dannypsnl** [comment]
@shana0440 It's ok now, just review it

File: response/response.go [hunk 1]
```diff
@@ -28,6 +28,12 @@ func New(body interface{}) *Response {
 }
 
 func (res *Response) Status(code int) *Response {
+	if !validStatusCode[code] {
+		panic(fmt.Errorf("reject invalid status code: %d", code))
+	}
+	if res.statusCode != 0 {
+		panic("status code already be changed, you can't modify it twice!")
```
Comment (shana0440) [inline response/response.go]: `status code already been set` is enough

File: response/response.go [hunk 2]
```diff
@@ -28,6 +28,12 @@ func New(body interface{}) *Response {
 }
 
 func (res *Response) Status(code int) *Response {
+	if !validStatusCode[code] {
```
Comment (shana0440) [inline response/response.go]: I don't think we should check status code is valid or invalid, maybe the user wants to use the custom status code.

File: response/valid_status_code_old.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
+// +build !go1.11
+
+package response
+
+import (
+	"net/http"
+)
+
+var validStatusCode = map[int]bool{
```
Comment (shana0440) [inline response/valid_status_code_old.go]: Why we have two `validStatusCode`?

File: response/response.go (hunk 2 above)
Comment (shana0440) [reply response/response.go]: But HTTP status codes are extensible.
From RFC 2616 https://tools.ietf.org/html/rfc2616#section-6.1.1

File: response/valid_status_code_old.go (hunk 3 above)
Comment (dannypsnl) [reply response/valid_status_code_old.go]: > The Status-Code element is a 3-digit integer
According to this, I think we should check it's `>= 100 && <= 999`?

File: response/response.go (hunk 2 above)
Comment (dannypsnl) [reply response/response.go]: > The Status-Code element is a 3-digit integer

According to this, I think we should check it's `>= 100 && <= 999`?

File: response/valid_status_code_old.go (hunk 3 above)
Comment (shana0440) [reply response/valid_status_code_old.go]: Yes, I think it will be great.

File: response/response.go (hunk 2 above)
Comment (shana0440) [reply response/response.go]: Yes, I think it will be great.

**Pull request merged**